src/ray/common/asio/instrumented_io_context.cc

```diff
@@ -19,31 +19,42 @@
 #include <iostream>
 #include <utility>
 
-/// C++11 helper for generalized lambda capture.
-template <typename T, typename F>
-class capture_impl {
-  T x;
-  F f;
+namespace {
 
- public:
-  capture_impl(T &&x, F &&f) : x{std::forward<T>(x)}, f{std::forward<F>(f)} {}
+/// A helper for creating a snapshot view of the global stats.
+/// This acquires a reader lock on the provided global stats, and creates a
+/// lockless copy of the stats.
+GlobalStats to_global_stats_view(std::shared_ptr<GuardedGlobalStats> stats) {
+  absl::MutexLock lock(&(stats->mutex));
+  return GlobalStats(stats->stats);
+}
 
-  template <typename... Ts>
-  auto operator()(Ts &&... args) -> decltype(f(x, std::forward<Ts>(args)...)) {
-    return f(x, std::forward<Ts>(args)...);
-  }
+/// A helper for creating a snapshot view of the stats for a handler.
+/// This acquires a lock on the provided guarded handler stats, and creates a
+/// lockless copy of the stats.
+HandlerStats to_handler_stats_view(std::shared_ptr<GuardedHandlerStats> stats) {
+  absl::MutexLock lock(&(stats->mutex));
+  return HandlerStats(stats->stats);
+}
 
-  template <typename... Ts>
-  auto operator()(Ts &&... args) const -> decltype(f(x, std::forward<Ts>(args)...)) {
-    return f(x, std::forward<Ts>(args)...);
-  }
-};
+/// A helper for converting a duration into a human readable string, such as "5.346 ms".
+std::string to_human_readable(double duration) {
+  static const std::array<std::string, 4> to_unit{{"ns", "us", "ms", "s"}};
+  size_t idx = std::min(to_unit.size() - 1,
+                        static_cast<size_t>(std::log(duration) / std::log(1000)));
+  double new_duration = duration / std::pow(1000, idx);
+  std::stringstream result;
+  result << std::fixed << std::setprecision(3) << new_duration << " " << to_unit[idx];
+  return result.str();
+}
 
-template <typename T, typename F>
-capture_impl<T, F> capture(T &&x, F &&f) {
-  return capture_impl<T, F>(std::forward<T>(x), std::forward<F>(f));
+/// A helper for converting a duration into a human readable string, such as "5.346 ms".
+std::string to_human_readable(int64_t duration) {
+  return to_human_readable(static_cast<double>(duration));
 }
 
+}  // namespace
+
 void instrumented_io_context::post(std::function<void()> handler,
```

Comment:
I just notice `instrumented_io_context` is not put in ray namespace. Do you think it's good to add `namespace ray`?

Reply:
Yeah I considered doing that in the first PR, but I noticed that some similar Ray utilities aren't in the `ray` namespace. I'd be happy to put it in the `ray` namespace (I do think it should be in that namespace), but I'd prefer to do it in another PR since it will probably touch a lot of files.

Reply:
Sure, that sounds good to me